File: chess_app/lib/features/endgame_trainer/models/endgame_catalog.dart
/// What there is to practise, as the server counts it.
///
/// The collection holds 128 distinct endings and grows with every mining run,
/// so this is not a list anybody maintains: the server derives the families and
/// their Serbian names from the material key itself, and the app only draws
/// what it is handed.
///
/// The counts come broken down by rating band as well, which is what lets the
/// picker show an exact total for any combination before it asks for a
/// position. A picker that lets you build an empty query and only says so
/// afterwards is the thing this avoids.
library;

class EloBand {
  const EloBand({required this.id, required this.name});

  final String id;
  final String name;

  factory EloBand.fromJson(Map<String, dynamic> json) => EloBand(
        id: json['id']?.toString() ?? '',
        name: json['name']?.toString() ?? '',
      );
}

/// One exact ending: 'KRPvKR', "top i pešak protiv topa", and how many.
class EndgameEnding {
  const EndgameEnding({
    required this.material,
    required this.label,
    required this.count,
    this.bands = const {},
    this.opposite = 0,
  });

  final String material;
  final String label;
  final int count;

  /// How the positions split across the rating bands.
  final Map<String, int> bands;

  /// How many of them have bishops on opposite colours — which the material
  /// key cannot say, so the server reads it from the position.
  final int opposite;

  int countIn(String? bandId) => bandId == null ? count : (bands[bandId] ?? 0);

  factory EndgameEnding.fromJson(Map<String, dynamic> json) => EndgameEnding(
        material: json['material']?.toString() ?? '',
        label: json['label']?.toString() ?? '',
        count: (json['count'] as num?)?.toInt() ?? 0,
        bands: ((json['bands'] as Map?) ?? const {}).map(
          (key, value) =>
              MapEntry(key.toString(), (value as num?)?.toInt() ?? 0),
        ),
        opposite: (json['opposite'] as num?)?.toInt() ?? 0,
      );
}

/// A family of endings: rook endings, pawn endings, and five more.
class EndgameFamily {
  const EndgameFamily({
    required this.id,
    required this.name,
    required this.count,
    required this.endings,
  });

  final String id;
  final String name;
  final int count;
  final List<EndgameEnding> endings;

  factory EndgameFamily.fromJson(Map<String, dynamic> json) => EndgameFamily(
        id: json['id']?.toString() ?? '',
        name: json['name']?.toString() ?? '',
        count: (json['count'] as num?)?.toInt() ?? 0,
        endings: ((json['endings'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(EndgameEnding.fromJson)
            .toList(),
      );
}

class EndgameCatalog {
  const EndgameCatalog({
    required this.families,
    required this.bands,
    this.oppositeBishops = 0,
  });

  final List<EndgameFamily> families;
  final List<EloBand> bands;

  /// How many positions in the whole collection have opposite bishops. Zero
  /// means the switch is not worth offering.
  final int oppositeBishops;

  bool get isEmpty => families.isEmpty;

  Iterable<EndgameEnding> get allEndings =>
      families.expand((family) => family.endings);

  /// Every ending, selected by default: a picker that starts empty asks the
  /// reader to do work before they can do anything.
  Set<String> get allMaterials =>
      allEndings.map((ending) => ending.material).toSet();

  /// Exactly how many positions a given choice covers.
  ///
  /// [bandId] null means every level. [oppositeOnly] counts the opposite-bishop
  /// positions instead, and ignores the band — the server reports that number
  /// per ending rather than per ending and band, and 43 positions in the whole
  /// collection is not worth a second breakdown.
  int countFor({
    required Set<String> materials,
    String? bandId,
    bool oppositeOnly = false,
  }) {
    var total = 0;
    for (final ending in allEndings) {
      if (!materials.contains(ending.material)) continue;
      total += oppositeOnly ? ending.opposite : ending.countIn(bandId);
    }
    return total;
  }

  /// The filter for a selection, or null when it covers everything.
  String? materialsParamFor(Set<String> materials) {
    if (materials.length >= allMaterials.length) return null;
    final ordered = materials.toList()..sort();
    return ordered.join(',');
  }

  factory EndgameCatalog.fromJson(Map<String, dynamic> json) => EndgameCatalog(
        families: ((json['families'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(EndgameFamily.fromJson)
            .toList(),
        bands: ((json['bands'] as List?) ?? const [])
            .whereType<Map<String, dynamic>>()
            .map(EloBand.fromJson)
            .toList(),
        oppositeBishops: (json['oppositeBishops'] as num?)?.toInt() ?? 0,
      );
}

/// What the picker sends back: which endings, which level, and whether to keep
/// only the opposite-bishop ones.
///
/// [materialsParam] is already resolved, because only the picker holds the
/// catalog and only the catalog knows whether a selection is all of it. Working
/// that out anywhere further along would mean carrying the catalog there too.
class EndgameChoice {
  const EndgameChoice({
    required this.materials,
    this.materialsParam,
    this.bandId,
    this.oppositeOnly = false,
  });

  final Set<String> materials;

  /// Null when everything is selected: the route reads a missing filter as
  /// "any", and 128 keys in a URL would say the same thing the long way.
  final String? materialsParam;

  final String? bandId;
  final bool oppositeOnly;
}

File: chess_backend/services/endgameCatalog.js
// endgameCatalog.js — sorting material keys into something a trainer can pick.
//
// The mined collection carries 128 distinct endings and grows with every run,
// so the grouping is a rule rather than a table. A key is not a name: 'KRPPvKR'
// says white has a rook and two pawns and black has a rook, and both the family
// it belongs to and the Serbian sentence for it follow from that.
//
// Two things this got wrong on the first try, kept here because both are easy
// to walk back into:
//
// The key does not put the stronger side first. 'KPPvKR' is a rook against two
// pawns, with the rook second. Rules written as "white is the armed one" put a
// fifth of the collection in a leftovers bucket.
//
// And "protiv" takes the genitive, so the two sides are not spelled the same
// way: "top i pešak protiv topa i pešaka", not "protiv top i pešak".
//
// What the key cannot say is whether two bishops stand on the same colour, and
// opposite bishops are a different ending altogether. That is read from the
// position instead - see oppositeBishops below.

/// 'KRPPvKR' -> [{R:1,P:2}, {R:1}]. Kings are dropped; every ending has two.
function parseMaterial(key) {
  const text = String(key || '');
  const halves = text.split('v');
  if (halves.length !== 2) return null;
  const count = (side) => {
    const out = {};
    for (const ch of side) {
      if (ch === 'K') continue;
      if (!'QRBNP'.includes(ch)) return null;
      out[ch] = (out[ch] || 0) + 1;
    }
    return out;
  };
  const white = count(halves[0]);
  const black = count(halves[1]);
  if (white === null || black === null) return null;
  return [white, black];
}

const minors = (s) => (s.B || 0) + (s.N || 0);
const pieces = (s) => (s.Q || 0) + (s.R || 0) + minors(s);

const FAMILIES = [
  { id: 'rooks', name: 'Topovske završnice' },
  { id: 'rook_vs_minors', name: 'Top protiv lakih figura' },
  { id: 'queens', name: 'Damske završnice' },
  { id: 'queen_vs_rest', name: 'Dama protiv ostalog materijala' },
  { id: 'minors', name: 'Lake figure' },
  { id: 'pawns', name: 'Pešačke završnice' },
  { id: 'pawns_vs_pieces', name: 'Pešaci protiv figura' },
];

const FAMILY_NAMES = Object.fromEntries(FAMILIES.map((f) => [f.id, f.name]));

/// Which family an ending belongs to, or null when the key is not one.
///
/// A rook apiece stays a rook ending even with a bishop standing next to it:
/// R+B vs R is 360 positions of the collection and a trainer asking for rook
/// endings means it too.
function familyOf(key) {
  const parsed = parseMaterial(key);
  if (parsed === null) return null;
  const [a, b] = parsed;

  const armed = [a, b].filter((s) => pieces(s) > 0);
  if (armed.length === 0) return 'pawns';
  // One side has nothing but pawns. Which piece the other side holds is the
  // smaller question - the exercise is the same shape either way.
  if (armed.length === 1) return 'pawns_vs_pieces';

  if ((a.Q || 0) > 0 && (b.Q || 0) > 0) return 'queens';
  if ((a.Q || 0) > 0 || (b.Q || 0) > 0) return 'queen_vs_rest';
  if ((a.R || 0) > 0 && (b.R || 0) > 0) return 'rooks';
  if ((a.R || 0) > 0 || (b.R || 0) > 0) return 'rook_vs_minors';
  return 'minors';
}

// Nominative for the side that acts, genitive for the side after "protiv".
const ONE = { Q: 'dama', R: 'top', B: 'lovac', N: 'skakač', P: 'pešak' };
const ONE_GEN = { Q: 'dame', R: 'topa', B: 'lovca', N: 'skakača', P: 'pešaka' };
const MANY = { Q: 'dame', R: 'topa', B: 'lovca', N: 'skakača', P: 'pešaka' };
const HOW = {
  2: 'dva', 3: 'tri', 4: 'četiri', 5: 'pet', 6: 'šest', 7: 'sedam', 8: 'osam',
};
const ORDER = ['Q', 'R', 'B', 'N', 'P'];

function describeSide(set, genitive) {
  const parts = [];
  for (const piece of ORDER) {
    const n = set[piece] || 0;
    if (!n) continue;
    parts.push(n === 1
      ? (genitive ? ONE_GEN[piece] : ONE[piece])
      : `${HOW[n] || n} ${MANY[piece]}`);
  }
  if (parts.length === 0) return genitive ? 'golog kralja' : 'goli kralj';
  if (parts.length === 1) return parts[0];
  return `${parts.slice(0, -1).join(', ')} i ${parts[parts.length - 1]}`;
}

/// 'KRPPvKR' -> 'top i dva pešaka protiv topa'.
function labelOf(key) {
  const parsed = parseMaterial(key);
  if (parsed === null) return String(key || '');
  return `${describeSide(parsed[0], false)} protiv ${describeSide(parsed[1], true)}`;
}

/// Whether the two bishops stand on squares of different colours.
///
/// Not in the key and not derivable from it: 'KBPvKBP' covers both the ending
/// where a extra pawn usually wins and the one where it usually does not. Null
/// where the question does not arise - which is most of the collection - so a
/// filter can tell "no" from "not applicable".
function oppositeBishops(fen) {
  const board = String(fen || '').split(' ')[0];
  if (!board) return null;
  const squares = { white: [], black: [] };
  let file = 0;
  let rank = 7;
  for (const ch of board) {
    if (ch === '/') {
      rank -= 1;
      file = 0;
      continue;
    }
    if (ch >= '1' && ch <= '8') {
      file += Number(ch);
      continue;
    }
    if (ch === 'B') squares.white.push((file + rank) % 2);
    if (ch === 'b') squares.black.push((file + rank) % 2);
    file += 1;
  }
  if (squares.white.length !== 1 || squares.black.length !== 1) return null;
  return squares.white[0] !== squares.black[0];
}

/// The rating bands the picker offers.
///
/// Only positions taken from real mistakes carry a rating - the mined ones were
/// found by measurement, not by anyone failing - so 'mined' is a band of its
/// own rather than a gap. A trainer picking a level is picking the level
/// somebody actually got it wrong at, and that is worth keeping separate from
/// "we do not know".
const ELO_BANDS = [
  { id: 'mined', name: 'Bez rejtinga (izrudareno)', min: null, max: null },
  { id: 'u1800', name: 'Do 1800', min: 0, max: 1799 },
  { id: 'b1800', name: '1800 - 2000', min: 1800, max: 1999 },
  { id: 'b2000', name: '2000 - 2200', min: 2000, max: 2199 },
  { id: 'b2200', name: '2200 - 2400', min: 2200, max: 2399 },
  { id: 'b2400', name: '2400 i preko', min: 2400, max: 9999 },
];

/// The SQL that puts a row in one of them. Kept next to the bands so the two
/// cannot drift: a band whose bounds live in one file and whose CASE lives in
/// another is a bug waiting for the first edit.
const ELO_BAND_SQL = `CASE
  WHEN blunder_elo IS NULL THEN 'mined'
  WHEN blunder_elo < 1800 THEN 'u1800'
  WHEN blunder_elo < 2000 THEN 'b1800'
  WHEN blunder_elo < 2200 THEN 'b2000'
  WHEN blunder_elo < 2400 THEN 'b2200'
  ELSE 'b2400'
END`;

/// Groups counted material keys into families, biggest family first and the
/// biggest ending inside each.
///
/// Rows may carry a `band`, in which case each ending also reports how its
/// positions split across the bands - which is what lets the picker show a
/// total for any combination of endings and levels without asking again.
function buildCatalog(rows) {
  const byFamily = new Map();
  const endings = new Map();
  for (const row of rows) {
    const family = familyOf(row.material);
    if (family === null) continue;
    const count = Number(row.n) || 0;

    if (!byFamily.has(family)) byFamily.set(family, { total: 0, keys: [] });
    const bucket = byFamily.get(family);
    bucket.total += count;

    // One ending can arrive as several rows, one per band.
    let ending = endings.get(row.material);
    if (!ending) {
      ending = {
        material: row.material,
        label: labelOf(row.material),
        count: 0,
        bands: {},
        opposite: 0,
      };
      endings.set(row.material, ending);
      bucket.keys.push(ending);
    }
    ending.count += count;
    ending.opposite += Number(row.opposite) || 0;
    if (row.band) ending.bands[row.band] = (ending.bands[row.band] || 0) + count;
  }

  return [...byFamily.entries()]
    .map(([id, bucket]) => ({
      id,
      name: FAMILY_NAMES[id] || id,
      count: bucket.total,
      endings: bucket.keys.sort((x, y) => y.count - x.count),
    }))
    .sort((x, y) => y.count - x.count);
}

module.exports = {
  ELO_BANDS,
  ELO_BAND_SQL,
  parseMaterial,
  familyOf,
  labelOf,
  oppositeBishops,
  buildCatalog,
  FAMILIES,
};
